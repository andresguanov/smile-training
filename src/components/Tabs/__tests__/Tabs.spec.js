import { shallowMount, mount } from "@vue/test-utils"
import smTabs from "@components/Tabs/Tabs"
import smTab from "@components/Tabs/Tab"
import smTabsBar from "@components/Tabs/TabsBar"
import smTabItem from "@components/Tabs/TabItem"
import smTabsItems from "@components/Tabs/TabsItems"

let wrapper

describe("smTabs", () => {
  beforeEach(() => {
    wrapper = mount(smTabs, {
      propsData: {
        value: ""
      },
      slots: {
        default: [
          "<sm-tab id='1' > 1 </sm-tab>",
          "<sm-tab id='2' > 2 </sm-tab>",
          "<sm-tab id='3' > 3 </sm-tab>",
          "<sm-tab-item value='1' > 1 </sm-tab-item>",
          "<sm-tab-item value='2' > 2 </sm-tab-item>",
          "<sm-tab-item value='3' > 3 </sm-tab-item>"
        ]
      },
      stubs: {
        "sm-tab": smTab,
        "sm-tab-item": smTabItem
      }
    })
  })
  //#region Basic render
  it("renders default smTabs", () => {
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.findComponent(smTabsBar).exists()).toBe(true)
    expect(wrapper.findComponent(smTabsItems).exists()).toBe(true)
  })
  //#endregion

  //#region Props
  it("accepts only valid tabsPositions", async () => {
    const validator = wrapper.vm.$options.props.tabsPosition.validator

    expect(validator("left")).toBe(true)
    expect(validator("right")).toBe(true)
    expect(validator("center")).toBe(true)
    expect(validator("top")).toBe(false)
    expect(validator("bottom")).toBe(false)
    expect(validator("other")).toBe(false)

    await wrapper.setProps({ tabsPosition: "right" })

    expect(wrapper.find(".sm-tabs-bar.right").exists()).toBe(true)
  })
  it("grows", async () => {
    await wrapper.setProps({ grow: true })

    expect(wrapper.find(".sm-tabs-bar.grow").exists()).toBe(true)
  })
  //#endregion

  //#region Methods
  it("changes", () => {
    let value = ""
    const input = jest.fn(v => (value = v))
    wrapper = mount(smTabs, {
      propsData: {
        value
      },
      slots: {
        default: [
          "<sm-tab id='1' > 1 </sm-tab>",
          "<sm-tab id='2' > 2 </sm-tab>",
          "<sm-tab id='3' > 3 </sm-tab>",
          "<sm-tab-item value='1' > 1 </sm-tab-item>",
          "<sm-tab-item value='2' > 2 </sm-tab-item>",
          "<sm-tab-item value='3' > 3 </sm-tab-item>"
        ]
      },
      stubs: {
        "sm-tab": smTab,
        "sm-tab-item": smTabItem
      },
      listeners: {
        input
      }
    })
    wrapper.vm.change("tab")
    expect(input).toHaveBeenCalledTimes(1)
    expect(input).toHaveBeenCalledWith("tab")
  })

  it("parses nodes", () => {
    const { tabs, items } = wrapper.vm.parseNodes()
    expect(tabs.filter(tab => tab.componentOptions.tag === "sm-tab").length).toBe(3)
    expect(items.filter(item => item.componentOptions.tag === "sm-tab-item").length).toBe(3)
  })
  //#endregion

  //#region Events
  //#endregion

  //#region Slots
  //#endregion
})
