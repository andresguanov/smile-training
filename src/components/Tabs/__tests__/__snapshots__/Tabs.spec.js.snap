// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smTabs renders default smTabs 1`] = `
<div class="sm-tabs">
  <div class="sm-tabs-bar left">
    <div class="sm-tab active" style="transform: translateX(0px);"><span> 1 </span></div>
    <div class="sm-tab" style="transform: translateX(0px);"><span> 2 </span></div>
    <div class="sm-tab" style="transform: translateX(0px);"><span> 3 </span></div>
  </div>
  <div class="sm-tabs-items">
    <div class="sm-tab-item active"> 1 </div>
    <div class="sm-tab-item"> 2 </div>
    <div class="sm-tab-item"> 3 </div>
  </div>
</div>
`;
